pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,373 @@
+	}
+	cliArgs := fmt.Sprintf("--connect '%s' --trustall --file '%s/%s'", url, BatchVolumeRoot, BatchFilename)
+
+	labels := batchLabels(batch.Name)
+	infinispan.AddLabelsForPods(labels)
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      batch.Name,
+			Namespace: batch.Namespace,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: pointer.Int32Ptr(0),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:    batch.Name,
+						Image:   infinispan.ImageName(),
+						Command: []string{"/opt/infinispan/bin/cli.sh", cliArgs},
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      BatchVolumeName,
+								MountPath: BatchVolumeRoot,
+							},
+						},
+					}},
+					RestartPolicy: corev1.RestartPolicyNever,
+					Volumes: []corev1.Volume{{
+						Name: BatchVolumeName,
+						VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{
+								LocalObjectReference: corev1.LocalObjectReference{Name: *batch.Spec.ConfigMap},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, job, func() error {
+		return controllerutil.SetControllerReference(batch, job, r.scheme)
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create batch job '%s': %w", batch.Name, err)
+	}
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchRunning, nil)
+}
+
+func (r *batchResource) waitToComplete() (reconcile.Result, error) {
+	batch := r.instance
+	job := &batchv1.Job{}
+	if result, err := kube.LookupResource(batch.Name, batch.Namespace, job, r.client, log); result != nil {
+		return *result, err
+	}
+
+	status := job.Status
+	if status.Succeeded > 0 {
+		return reconcile.Result{}, r.UpdatePhase(v2.BatchSucceeded, nil)
+	}
+
+	if status.Failed > 0 {
+		numConditions := len(status.Conditions)
+		if numConditions > 0 {
+			condition := status.Conditions[numConditions-1]
+
+			if condition.Type == batchv1.JobFailed {
+				var reason string
+				podName, err := r.getJobPodName(batch.Name, batch.Namespace)
+				if err != nil {
+					reason = err.Error()
+				} else {
+					reason, err = kubernetes.Logs(podName, batch.Namespace)
+					if err != nil {
+						reason = fmt.Errorf("Unable to retrive logs for batch %s: %w", batch.Name, err).Error()
+					}
+				}
+
+				_, err = r.update(func() error {
+					r.instance.Status.Phase = v2.BatchFailed
+					r.instance.Status.Reason = reason
+					return nil
+				})
+				return reconcile.Result{}, err
+			}
+		}
+	}
+	// The job has not completed, wait 1 second before retrying
+	return reconcile.Result{}, nil
+}
+
+func (r *batchResource) cleanup() (reconcile.Result, error) {
+	batch := r.instance
+	meta := metav1.ObjectMeta{
+		Namespace: batch.Namespace,
+		Name:      batch.Name,
+	}
+	var allErrors error
+	// Delete the Job used to execute the batch
```

Comment:
It would be nice to keep Job and Job Pods undeleted and cleanup them on Batch CR garbage collection (delete)
If some issues will occurs we will need them for debug and trace

Reply:
Updated as a new commit. Let me know what you think. I'll rebase if you're happy with the changes.

Reply:
Thanks, @ryanemerson , works excelent

Reply:
Great. I've rebased to squash the commit.